scientist.go

```diff
@@ -54,45 +62,91 @@ func Run(e *Experiment, name string) Result {
 		r.Errors = append(r.Errors, e.resultErr("before_run", err))
 	}
 
-	numCandidates := len(e.behaviors) - 1
-	r.Control = observe(e, name, e.behaviors[name])
-	r.Candidates = make([]*Observation, numCandidates)
-	r.Ignored = make([]*Observation, 0, numCandidates)
-	r.Mismatched = make([]*Observation, 0, numCandidates)
-	r.Observations = make([]*Observation, numCandidates+1)
-	r.Observations[0] = r.Control
-
-	i := 0
-	for bname, b := range e.behaviors {
-		if bname == name {
-			continue
-		}
-
-		c := observe(e, bname, b)
-		r.Candidates[i] = c
-		i += 1
-		r.Observations[i] = c
-
-		ok, err := matching(e, r.Control, c)
-		if err != nil {
-			ok = false
-			r.Errors = append(r.Errors, e.resultErr("compare", err))
+	numBehaviors := len(e.behaviors)
+
+	r.Candidates = make([]*Observation, numBehaviors-1)
+	r.Ignored = make([]*Observation, 0, numBehaviors-1)
+	r.Mismatched = make([]*Observation, 0, numBehaviors-1)
+	r.Observations = make([]*Observation, numBehaviors)
+
+	if !e.runConcurrently {
+		r.Control = observe(e, name, e.behaviors[name])
+		r.Observations[0] = r.Control
+
+		i := 0
+		for bname, b := range e.behaviors {
+			if bname == name {
+				continue
+			}
+			c := observe(e, bname, b)
+			r.Candidates[i] = c
+			i += 1
+			r.Observations[i] = c
+			processObservation(e, &r, r.Control, c)
 		}
-
-		if ok {
-			continue
+	} else {
+		resultChan := make(chan observationResult, numBehaviors)
+		var wg sync.WaitGroup
+
+		for bname, b := range e.behaviors {
+			wg.Add(1)
+			go func(behaviorName string, behavior behaviorFunc) {
+				defer wg.Done()
+
+				var ctx context.Context
+				var cancel context.CancelFunc
+
+				ctx, cancel = createContext(e)
+				defer cancel()
+
+				doneChan := make(chan *Observation, 1)
+				go func() {
+					obs := observe(e, behaviorName, behavior)
+					doneChan <- obs
+				}()
+
+				select {
+				case obs := <-doneChan:
+					resultChan <- observationResult{behaviorName, obs, nil}
+				case <-ctx.Done():
+					timeoutErr := ctx.Err()
+					resultChan <- observationResult{
+						name: behaviorName,
+						obs: &Observation{
+							Name: behaviorName,
+						},
+						err: timeoutErr,
+					}
+				}
+			}(bname, b)
 		}
 
-		ignored, err := ignoring(e, r.Control, c)
-		if err != nil {
-			ignored = false
-			r.Errors = append(r.Errors, e.resultErr("ignore", err))
+		go func() {
+			wg.Wait()
+			close(resultChan)
+		}()
+
+		i := 0
+		for res := range resultChan {
```

Comment:
does this block until channel closure?

Reply:
valid point, the wg.wait and channel closure could be w/o go func this will ensure no race between iterating the channel and closing it

Reply:
Extracted wait from gourotine